plugins/module_utils/cmci.py

```diff
@@ -603,67 +587,153 @@ def append_attributes(self, element):
+                    "resources -> complex_filter%s" % (type(i), path)
+                )
+
+            valid_keys = [AND, ATTRIBUTE, OPERATOR, OR, VALUE]
+            diff = set(i.keys()) - set(valid_keys)
+            if len(diff) != 0:
+                self._fail(
+                    "Unsupported parameters for (basic.py) module: %s found"
+                    " in resources -> complex_filter%s. Supported parameters "
+                    "include: %s"
+                    % (", ".join(diff), path, ", ".join(valid_keys))
+                )
 
+            # Validate required_one_of
+            and_item = i.get(AND)
+            or_item = i.get(OR)
+            attribute = i.get(ATTRIBUTE)
 
-def _convert_filter_operator(operator):
-    if operator in ['<', 'LT']:
-        return '<'
-    if operator in ['<=', 'LE']:
-        return '<='
-    if operator in ['=', 'EQ']:
-        return '='
-    if operator in ['>=', 'GE']:
-        return '>='
-    if operator in ['>', 'GT']:
-        return '>'
-    if operator in ['¬=', '!=', 'NE']:
-        return '¬='
-    if operator in ['==', 'IS']:
-        return '=='
-
-
-def _get_filter(list_of_filters, complex_filter_string, joiner):
-    for i in list_of_filters:
-        and_item = i.get('and')
-        or_item = i.get('or')
-        attribute = i.get('attribute')
-
-        if and_item is not None:
-            and_filter_string = _get_filter(and_item, '', ' AND ')
-            complex_filter_string = _append_filter_string(
-                complex_filter_string,
-                and_filter_string, joiner
-            )
-        if or_item is not None:
-            or_filter_string = _get_filter(or_item, '', ' OR ')
-            complex_filter_string = _append_filter_string(
-                complex_filter_string,
-                or_filter_string, joiner
-            )
-        if attribute is not None:
-            operator = _convert_filter_operator(i.get('operator'))
-            value = i.get('value')
-
-            if operator == '¬=':
-                # Provides a filter string in the format NOT(FOO=='BAR')
-                attribute_filter_string =\
-                    'NOT(' + attribute + '==' + '\'' + value + '\'' + ')'
-            else:
-                attribute_filter_string = \
-                    attribute + operator + '\'' + value + '\''
+            if not and_item and not or_item and not attribute:
+                self._fail(
+                    "one of the following is required: %s found"
+                    " in resources -> complex_filter%s"
+                    % (", ".join([ATTRIBUTE, AND, OR]), path)
+                )
 
-            complex_filter_string = _append_filter_string(
-                complex_filter_string,
-                attribute_filter_string,
-                joiner
-            )
+            # Validate required_by
+            op = i.get(OPERATOR)
+
+            if op and not attribute:
```

Comment:
Do we need to do the same for a value with no attribute?

Reply:
Yes, I think you're right... let me write a test and see if something weird happens.  My go to...